query.go

```diff
@@ -0,0 +1,431 @@
+package jsonmap
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"reflect"
+	"regexp"
+	"strconv"
+	"time"
+)
+
+// This is the overarching struct used to transform structs into url params
+// and vice versa
+type QueryMap struct {
+	UnderlyingType interface{}
+	Parameters     []MappedParameter
+}
+
+// Taking a struct and turning it into a url param. The precise mechanisms of doing
+// so are are defined in the individual MappedParameter
+func (qm QueryMap) Encode(src interface{}, urlQuery map[string][]string) error {
+	srcVal := reflect.ValueOf(src)
+
+	for _, p := range qm.Parameters {
+		fieldVal := srcVal.FieldByName(p.StructFieldName)
+
+		if fieldVal.IsZero() && p.OmitEmpty {
+			continue
+		}
+
+		strVal, err := p.Mapper.Encode(fieldVal)
+		if err != nil {
+			return errors.New("error in encoding struct: " + err.Error())
+		}
+
+		urlQuery[p.ParameterName] = strVal
+	}
+
+	return nil
+}
+
+// Taking a URL Query (or any string->[]string struct) and shoving it into the struct
+// as specified by qm.UnderlyingType
+func (qm QueryMap) Decode(urlQuery map[string][]string, dst interface{}) error {
+	// First sanity check to ensure that the struct passed in matches
+	// the struct the QueryMap was designed to handle
+	if reflect.ValueOf(dst).Elem().Type() != reflect.TypeOf(qm.UnderlyingType) {
+		return fmt.Errorf("attempting to decode into mismatched struct: expected %s but got %s",
+			reflect.TypeOf(qm.UnderlyingType),
+			reflect.ValueOf(dst).Elem().Type(),
+		)
+	}
+
+	errs := &MultiValidationError{}
+	dstVal := reflect.ValueOf(dst).Elem()
+	for _, param := range qm.Parameters {
+		paramValue := urlQuery[param.ParameterName]
```

Comment:
Do we care about query params set with no value? I can imagine something like `?bar&foo=jirwin&foo=matcha` resulting in a map that looks like:
```
map[string][]string{
  "bar": [],
  "foo": ["jirwin", "matcha"],
}
```

As it stands, the code will ignore that `bar` was set. Maybe a query param set without a value should be decoded as `true` bool?

Reply:
Yeah, this logic should prob be moved down to `MappedParameter` so it can handle empty parameters on a more granular basis.